doc/sr_integration.rst

```diff
@@ -0,0 +1,211 @@
+
+1. Find locations where file I/O would normally happen and either replace
+   or add code to use the SmartRedis client and store the data in the
+   database
+2. Use the `key_prefix` created during initialization to avoid accidental
+   writes/reads from different clients
+3. Use the SmartSim `dataset` type when using clients representing decomposed
+   subdomains to make the retrieval/use of the data more performant
+
+============
+Full example
+============
+
+The following pseudocode is used to demonstrate various aspects of instrumenting an
+existing simulation code with SmartRedis. This code is representative of solving
+the time-evolving heat equation. but will be augmented using an ML model to
+provide a preconditioning step each iteration and post the state of the simulation
+to the database. ::
+
+    program main
+
+        ! Initialize the model, setup MPI, communications, read input files
+        call initialize_model( temperature, number_of_timesteps )
+
+        main_loop: do i=1,number_of_timesteps
+
+            ! Write the current state of the simulation to a file
+            call write_current_state(temperature)
+
+            ! Call a time integrator to step the temperature field forward
+            call timestep_simulation(temperature)
+
+        enddo
+    end program main
+
+Following the guidelines from above, the first step is to initialize the client
+and create a unique identifier for the given processor. This should be done
+within roughly the same portion of the code where the rest of the model. ::
+
+    ! Import SmartRedis modules
+    use, only smartredis_client : client_type
+
+    ! Declare a new variable called client and a string to create a unique
+    ! name for keys
+    type(client_type) :: smartredis_client
+    character(len=7)  :: key_prefix
+    integer :: mpi_rank, mpi_code, smartredis_code
+
+    ! Note adding use_cluster as an additional runtime argument for SmartRedis
+    call initialize_model(temperature, number_of_timesteps, use_cluster)
+    call smartredis_client%initialize(use_cluster)
+    call MPI_Comm_rank(MPI_COMM_WORLD, mpi_rank, mpi_code)
+    write(key_prefix,'(I6.6,A)') mpi_rank, '_'
+
+    ! Assume all ranks will use the same machine learning model
+    if (mpi_rank==0) call set_model_from_file("example_model_key", "path/to/model.pt", "TORCH", "gpu")
+
+Next, add the calls in the main loop to send the temperature to the database ::
+
+    character(len=10), dimension(1) :: model_input, model_output
+
+    main_loop: do i=1,number_of_timesteps
+
+        ! Write the current state of the simulation to a file
+        call write_current_state(temperature)
+        model_input(1) = key_prefix//"temperature"
+        model_output(1) = key_prefix//"temperature_out"
+        call smartredis_client%put_tensor(model_input(1))
+
+        ! Run the machine learning model
+        return_code = smartredis_client%run_model("example_model_key", model_input, model_output)
```

Comment:
Call out that we don't need to prefix the model key?

Reply:
Good suggestion